### PR TITLE
feat: Store query method

### LINKED ISSUE #2220: Query method for storage
This will be needed for the upcoming encrypted storage functionality.

(it's also just a useful thing to have in general, really)

### PR DESCRIPTION
**Pull request by DRK3**
- Added CouchDB implementation
- Other implementations are left as stubs for now

closes #2220

Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

### CODE REVIEW

File: pkg/didcomm/protocol/didexchange/service_test.go [hunk 1]
```diff
@@ -733,6 +733,10 @@ func (m *mockStore) Delete(k string) error {
 	return m.delete(k)
 }
 
+func (m *mockStore) Query(query string) (storage.StoreIterator, error) {
+	panic("implement me")
```
Comment (sudeshrshetty) [inline pkg/didcomm/protocol/didexchange/service_test.go:737]: why panic??

Comment (sudeshrshetty) [reply pkg/didcomm/protocol/didexchange/service_test.go:737]: add a TODO if implementation is pending

Comment (sudeshrshetty) [reply pkg/didcomm/protocol/didexchange/service_test.go:737]: or throw error

File: pkg/storage/couchdb/couchdbstore_test.go [hunk 2]
```diff
@@ -290,6 +297,110 @@ func TestCouchDBStore(t *testing.T) {
...
+		   },
+			"use_index": ["` + testDesignDoc + `", "` + testIndexName + `"]
+		}`)
+			require.NoError(t, err)
+
+			ok = itr.Next()
+			require.True(t, ok)
+
+			value := itr.Value()
+			require.NoError(t, err)
```
Comment (sandrask) [inline pkg/storage/couchdb/couchdbstore_test.go:378]: Which err are you checking here? There are few references to same err bellow ...

Comment (DRK3) [reply pkg/storage/couchdb/couchdbstore_test.go]: Oops. Good catch - those weren't supposed to be there. Fixed!

File: pkg/storage/couchdb/couchdbstore.go [hunk 3]
```diff
@@ -303,14 +308,39 @@ func (c *CouchDBStore) Iterator(startKey, endKey string) storage.StoreIterator {
...
+	nextCallResult := i.resultRows.Next()
+
+	// Kivik only guarantees that this value will be set after all the rows have been iterated through.
+	warningMsg := i.resultRows.Warning()
+
+	if warningMsg != "" {
+		logger.Warnf(warningMsg)
+	}
+
+	if err := i.resultRows.Err(); err != nil {
```
Comment (sandrask) [inline pkg/storage/couchdb/couchdbstore.go:339]: Should this be here?

Comment (DRK3) [reply pkg/storage/couchdb/couchdbstore.go]: Nope! - Remove it.

File: pkg/storage/couchdb/couchdbstore_test.go [hunk 4]
```diff
@@ -290,6 +297,104 @@ func TestCouchDBStore(t *testing.T) {
...
+
+			itr, err := store.Query(`{
+		   "selector": {
+		       "payload.employeeID": 1234
+		   },
+			"use_index": ["` + testDesignDoc + `", "` + testIndexName + `"]
+		}`)
+			require.NoError(t, err)
+
+			ok = itr.Next()
```
Comment (sandrask) [inline pkg/storage/couchdb/couchdbstore_test.go:371]: Add error checks after next so developers can see how to use it properly

Comment (DRK3) [reply pkg/storage/couchdb/couchdbstore_test.go]: Done. I also refactored the tests a bit to reduce code duplication and make the two cases (query with a valid vs invalid index) more clear.

**Review by sandrask — APPROVED** [review]

**Review by sudeshrshetty — APPROVED** [review]

**Review by fqutishat — APPROVED** [review]

**Pull request merged**